File: frontend/src/types/game.ts
export type GameStatus = "waiting" | "playing" | "finished";

export interface Player {
  id: string;
  display_name: string;
  score: number;
  captured_ids: string[];
  is_connected: boolean;
}

export interface Game {
  id: string;
  room_code: string;
  status: GameStatus;
  deck_id: string;
  players: Player[];
  current_card_id: string;
  remaining_ids: string[];
  time_limit_sec: number;
}

// H4: removed "start" and "card_grabbed" that do not exist in backend ws/message.go
export type WSMessageType =
  | "join"
  | "card_revealed"
  | "grab"
  | "grab_result"
  | "game_over"
  | "player_joined"
  | "player_left"
  | "reconnect"
  | "match_found"
  | "waiting"
  | "game_state"
  | "error";

export interface WSMessage {
  type: WSMessageType;
  payload: unknown;
}

// --- Payload types matching backend ws/message.go ---

export interface JoinPayload {
  room_code: string;
  player_name: string;
  random_match: boolean;
}

export interface GrabPayload {
  card_id: string;
}

export interface ReconnectPayload {
  game_id: string;
  player_id: string;
}

export interface PlayerJoinedPayload {
  player_id: string;
  player_name: string;
  room_code: string;
}

export interface CandidateCard {
  id: string;
  name: string;
  image_url: string;
}

export interface CardRevealedPayload {
  reading_text: string;
  candidates: CandidateCard[];
  card_index: number;
  total_cards: number;
}

export interface GrabResultPayload {
  winner_id: string;
  winner_name: string;
  card_id: string;
  card_name: string;
  correct: boolean;
}

export interface PlayerResult {
  player_id: string;
  player_name: string;
  score: number;
  captured_ids: string[];
}

export interface GameOverPayload {
  players: PlayerResult[];
  winner_id: string;
}

export interface WaitingPayload {
  message: string;
}

export interface ErrorPayload {
  message: string;
}

export interface GameStatePayload {
  game_id: string;
  current_index: number;
  total_cards: number;
  status: GameStatus;
}

File: backend/internal/model/game.go
package model

import "time"

// GameStatus represents the current state of a game.
type GameStatus string

const (
	GameStatusWaiting  GameStatus = "waiting"
	GameStatusPlaying  GameStatus = "playing"
	GameStatusFinished GameStatus = "finished"
)

// Game represents an active karuta game session.
type Game struct {
	ID            string     `json:"id"`
	RoomCode      string     `json:"room_code"`
	Status        GameStatus `json:"status"`
	DeckID        string     `json:"deck_id"`
	Players       []Player   `json:"players"`
	CurrentCardID string     `json:"current_card_id"`
	RemainingIDs  []string   `json:"remaining_ids"`
	TimeLimitSec  int        `json:"time_limit_sec"`
	CreatedAt     time.Time  `json:"created_at"`
	StartedAt     *time.Time `json:"started_at,omitempty"`
	FinishedAt    *time.Time `json:"finished_at,omitempty"`
}

// Player represents a participant in a game.
type Player struct {
	ID           string   `json:"id"`
	DisplayName  string   `json:"display_name"`
	Score        int      `json:"score"`
	CapturedIDs  []string `json:"captured_ids"`
	IsConnected  bool     `json:"is_connected"`
	DisconnectAt *time.Time `json:"disconnect_at,omitempty"`
}
